core/src/main/scala/ox/resilience/AdaptiveRetry.scala

```
@@ -0,0 +1,132 @@
+  * @param successReward
+  *   Number of tokens to add back to [[tokenBucket]] after successful operation.
+  */
+case class AdaptiveRetry(
+    tokenBucket: TokenBucket,
+    failureCost: Int,
+    successReward: Int
+):
+  /** Retries an operation using the given error mode until it succeeds or the config decides to stop. Note that any exceptions thrown by
+    * the operation aren't caught (unless the operation catches them as part of its implementation) and don't cause a retry to happen.
+    *
+    * This is a special case of [[scheduledWithErrorMode]] with a given set of defaults. See [[RetryConfig]] for more details.
+    *
+    * @param config
+    *   The retry config - See [[RetryConfig]].
+    * @param shouldPayPenaltyCost
+    *   Function to decide if returned result [[T]] should be considered failure in terms of paying cost for retry. Penalty is paid only if
+    *   it is decided to retry operation, the penalty will not be paid for successful operation. Defaults to `true`.
+    * @param errorMode
+    *   The error mode to use, which specifies when a result value is considered success, and when a failure.
+    * @param operation
+    *   The operation to retry.
+    * @tparam E
+    *   type of error.
+    * @tparam T
+    *   type of result of an operation.
+    * @tparam F
+    *   the context inside which [[E]] or [[T]] are returned.
+    * @return
+    *   Either:
+    *   - the result of the function if it eventually succeeds, in the context of `F`, as dictated by the error mode.
+    *   - the error `E` in context `F` as returned by the last attempt if the config decides to stop.
+    * @see
+    *   [[scheduledWithErrorMode]]
+    */
+  def retryWithErrorMode[E, T, F[_]](
+      config: RetryConfig[E, T],
+      shouldPayPenaltyCost: T => Boolean = (_: T) => true,
+      errorMode: ErrorMode[E, F]
+  )(operation: => F[T]): F[T] =
+    val shouldAttempt: Either[E, T] => Boolean = result =>
```

this implementation comes close, but we're still doing a side-effect in a predicate here
I think we'd need only one new callback, `afterAttempt`, parametrised with the result (E/T), attempt number, and a flag if the operation will be run again. There, we can either pay the reward or the penalty

---

But we still need somehow to check before attempt if we have enough tokens. So at some point in loop in schedule we need to run `tokenBucket.tryAcquire(failureCost)` to determine if we can retry. I don't know if I understand correctly, but if we only consider bucket in `afterAttempt` and get rid of side effects in predicate we will be adding and removing tokens, but never consider if we have enough to retry.

---

Hm yeah I see the problem. I guess it's better to aquire the "retry tokens" immediately after an attempt, and fail early, instead of aquiring them before an attempt (after sleeping). Some tokens might be return meanwhile, but that's an edge case it itself, and if the tokens are already scarce, I guess it won't matter much if we aquire the retry tokens immediately after a failed attempt. Is that also what the Amazon SDK does?

As for the callbacks - maybe we can have a single one, `afterAttempt: Either[E, T] => ScheduleContinue’`? Instead of all the `shouldContinueOnError` etc. The name then is not "callback-like", so it should be clear that it shouldn't be called multiple times (as a predicate can). The return value (a simple Yes/No enum) specifies the intent of the return value. WDYT?

---

Amazon SDK indeed first tries to acquire token then wait.

As for the callback I think this would simplify config. Might need to change some things and tests but I think this should work